### PR TITLE
feat(integrations): Gitlab installation flow

### PR DESCRIPTION
**Pull request by MeredithAnya**
Setting up the installation flow for Gitlab 

Adds in checkbox for SSL verification:

screen shot 2018-09-13 at 3 13 28 pm

### CODE REVIEW

File: src/sentry/integrations/gitlab/integration.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+        data = state['identity']['data']
+        oauth_data = self.get_oauth_data(data)
+        user = self.get_user_info(data['access_token'], state['installation_data'])
+        group = self.get_group_info(data['access_token'], state['installation_data'])
+        scopes = sorted(GitlabIdentityProvider.oauth_scopes)
+        base_url = state['installation_data']['url']
+
+        integration = {
+            'name': group['name'],
+            # probs should add on base url to make it unique like ghe
```
Comment (lauryndbrown) [inline src/sentry/integrations/gitlab/integration.py]: Yes! Is this something we can do right now? Or is it something we have to wait until we have access to the gitlab instance james is working on?

File: src/sentry/integrations/gitlab/integration.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+            'external_id': group['id'],
```
Comment (lauryndbrown) [inline src/sentry/integrations/gitlab/integration.py]: I feel like adding the base url (or something) will be most important here. due to the unique constraint.

File: src/sentry/templates/sentry/integrations/gitlab-config.html [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+{% block modal_header_signout %} {% endblock %}
+
+{% block title %} {% trans "Gitlab Setup" %} | {{ block.super }} {% endblock %}
+
+{% block main %}
+<h3>Gitlab Configuration</h3>
+
+<form action="" method="post" class="form-stacked">
+  {% csrf_token %}
+  <input type="hidden" name="provider" value="vsts" />
```
Comment (lauryndbrown) [inline src/sentry/templates/sentry/integrations/gitlab-config.html]: this has `'vsts'` as the provider

Comment (lauryndbrown) [reply src/sentry/templates/sentry/integrations/gitlab-config.html]: Honestly, we could probably just extend the vsts template (create a base one and give it input), but it really isn't that bad.

**Review by lauryndbrown — DISMISSED** [review]
Looks great. You mostly need to add the default identity and test it out with enterprise edition.

File: src/sentry/integrations/gitlab/integration.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+from sentry.web.helpers import render_to_response
+from sentry.identity.pipeline import IdentityProviderPipeline
+from sentry.identity.gitlab.provider import GitlabIdentityProvider
+from sentry.integrations import Integration, IntegrationFeatures, IntegrationProvider, IntegrationMetadata
+from sentry.pipeline import NestedPipelineView, PipelineView
+from sentry.utils.http import absolute_uri
+
+
+DESCRIPTION = """
+Fill me out
```
Comment (lauryndbrown) [inline src/sentry/integrations/gitlab/integration.py]: I think just copying bb is fine here: 
https://github.com/getsentry/sentry/blob/master/src/sentry/integrations/bitbucket/integration.py#L18

File: src/sentry/templates/sentry/integrations/gitlab-config.html (hunk 3 above)
Comment (MeredithAnya) [reply src/sentry/templates/sentry/integrations/gitlab-config.html]: also it's funny cause I copy/pasta this from https://github.com/getsentry/sentry/blob/master/src/sentry/templates/sentry/integrations/github-enterprise-config.html which also has `vsts` 🙄

File: src/sentry/integrations/gitlab/integration.py [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+        """
+        Make the nested identity provider view. It is important that this view is
+        not constructed until we reach this step and the
+        ``oauth_config_information`` is available in the pipeline state. This
+        method should be late bound into the pipeline vies.
+        """
+        identity_pipeline_config = dict(
+            oauth_scopes=(),
+            redirect_url=absolute_uri('/extensions/gitlab/setup/'),
+            verify_ssl=False,
```
Comment (markstory) [inline src/sentry/integrations/gitlab/integration.py]: Why don't we want to verify ssl?

File: src/sentry/integrations/gitlab/integration.py [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+            verify=False
+        )
+
+        resp.raise_for_status()
+        return resp.json()
+
+    def get_group_info(self, access_token, installation_data):
+        session = http.build_session()
+        resp = session.get(
+            'https://{}/api/v4/groups/{}'.format(
```
Comment (markstory) [inline src/sentry/integrations/gitlab/integration.py]: You could have a UnicodeError here if the formatted data has non-ascii bytes.

**Comment by lauryndbrown** [comment]
@MeredithAnya Looking at the api, we need to have the user that installs the integration (the default identity) be an admin.

File: src/sentry/integrations/gitlab/integration.py (hunk 4 above)
Comment (MeredithAnya) [reply src/sentry/integrations/gitlab/integration.py]: im probs just going to leave this for now and make a jira ticket

File: src/sentry/integrations/gitlab/integration.py (hunk 5 above)
Comment (MeredithAnya) [reply src/sentry/integrations/gitlab/integration.py]: similar to GHE, for testing purposes. But should remember to remove this later 😬

File: src/sentry/integrations/gitlab/integration.py (hunk 4 above)
Comment (EvanPurkhiser) [reply src/sentry/integrations/gitlab/integration.py]: Leave it for now if you can, my feature stuff is coming changes some things around this anyway

File: src/sentry/integrations/gitlab/integration.py [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+
+class GitlabIntegration(Integration):
+
+    def get_client(self):
+        pass
+
+
+class InstallationForm(forms.Form):
+    url = forms.CharField(
+        label="Installation Url",
```
Comment (EvanPurkhiser) [inline src/sentry/integrations/gitlab/integration.py]: Label needs translated to no?

File: src/sentry/integrations/gitlab/integration.py [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+        help_text=_('The "base URL" for your github enterprise instance, '
+                    'includes the host and protocol.'),
+        widget=forms.TextInput(
+            attrs={'placeholder': _('https://github.example.com')}
```
Comment (EvanPurkhiser) [inline src/sentry/integrations/gitlab/integration.py]: This doesn't need translated (though looks like maybe this is wrong in GH too)

File: src/sentry/integrations/gitlab/integration.py (hunk 5 above)
Comment (MeredithAnya) [reply src/sentry/integrations/gitlab/integration.py]: @markstory actually added in a checkbox for the install form - will put up diff PR for GHE too

File: src/sentry/identity/gitlab/provider.py [hunk 9]
```diff
@@ -1,8 +1,24 @@
...
 
+def get_user_info(access_token, installation_data):
+    session = http.build_session()
+    resp = session.get(
+        u'https://{}/api/v4/user'.format(installation_data['url']),
+        headers={
+            'Accept': 'application/json',
+            'Authorization': 'Bearer %s' % access_token,
+        },
+        verify=False
```
Comment (markstory) [inline src/sentry/identity/gitlab/provider.py]: This seems risky. Could we only disable SSL when sentry is in dev/test mode?

File: src/sentry/integrations/gitlab/integration.py [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+    def dispatch(self, request, pipeline):
+        form = InstallationForm(request.POST)
+        if form.is_valid():
+            form_data = form.cleaned_data
+            form_data['url'] = urlparse(form_data['url']).netloc
+
+            pipeline.bind_state('installation_data', form_data)
+
+            pipeline.bind_state('oauth_config_information', {
+                "access_token_url": u"https://{}/oauth/token".format(form_data.get('url')),
```
Comment (lauryndbrown) [inline src/sentry/integrations/gitlab/integration.py]: nit: Why are we using double quotes here suddenly?

File: src/sentry/integrations/gitlab/integration.py [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+                'Authorization': 'Bearer %s' % access_token,
+            },
+            verify=installation_data['verify_ssl']
+        )
+
+        resp.raise_for_status()
+        return resp.json()
+
+    def get_pipeline_views(self):
+        return [InstallationConfigView(), lambda: self._make_identity_pipeline_view()]
```
Comment (lauryndbrown) [inline src/sentry/integrations/gitlab/integration.py]: hmmm am I wrong to think `self._make_identity_pipeline_view` would work? Seems like `lambda` is unnecessary.

File: tests/sentry/integrations/gitlab/__init__.py [hunk 12]
```diff
@@ -0,0 +1 @@
+from __future__ import absolute_import
```
Comment (lauryndbrown) [inline tests/sentry/integrations/gitlab/__init__.py]: I think it'll need to look like this: https://github.com/getsentry/sentry/blob/master/src/sentry/integrations/github/__init__.py

**Comment by lauryndbrown** [comment]
@MeredithAnya I think it looks good @EvanPurkhiser and/or @markstory are you satisfied with the verify ssl things?

**Review by markstory — APPROVED** [review]
SSL handling looks good to me.

**Pull request merged**